### PR TITLE
pflash_handler: correct the image path in and keep it compatible with pc

### PR DESCRIPTION
**Pull request by PaulYuuu**
1. Correct the image path from DATA_DIR to current_data_dir
2. Keep `pc` machine type can use continue to use OVMF

ID: 1926800
Signed-off-by: Yihuang Yu <yihyu@redhat.com>

### CODE REVIEW

**Comment by PaulYuuu** [comment]
Hello @pevogam , here is the hotfix of #2777
I will continue to work on this tomorrow.

**Comment by pevogam** [comment]
> Hello @pevogam , here is the hotfix of #2777
> I will continue to work on this tomorrow.

So far so good, the image directory problem is already resolved by the PR in its current form.

**Comment by PaulYuuu** [comment]
> > Hello @pevogam , here is the hotfix of #2777
> > I will continue to work on this tomorrow.
> 
> So far so good, the image directory problem is already resolved by the PR in its current form.

Well, thank you for the testing.
Since tomorrow will be the Spring Festival holiday(Feb 11th- Feb 17th) in China, I am not sure we can merge it before it. If the test scenario is important to you, you can merge it and fix new problems if raised.

File: virttest/qemu_devices/qcontainer.py
Comment (PaulYuuu) [inline virttest/qemu_devices/qcontainer.py:1073]: Hello @pevogam @luckyh , what's your opinion about it?
Without `images_base_dir`, I prefer to put all images created by Avocado-vt under `DATA_DIR/images`, but now images can be provided if `images_base_dir` is defined by users, so let them in the same directory also make senses with me.

Comment (luckyh) [reply virttest/qemu_devices/qcontainer.py:1073]: +1, thanks.

Comment (pevogam) [reply virttest/qemu_devices/qcontainer.py:1073]: Yes, makes sense, +1.

**Comment by pevogam** [comment]
> > > Hello @pevogam , here is the hotfix of #2777
> > > I will continue to work on this tomorrow.
> > 
> > 
> > So far so good, the image directory problem is already resolved by the PR in its current form.
> 
> Well, thank you for the testing.
> Since tomorrow will be the Spring Festival holiday(Feb 11th- Feb 17th) in China, I am not sure we can merge it before it. If the test scenario is important to you, you can merge it and fix new problems if raised.

Hi @PaulYuuu, for now the base image directory works, we will have to take a look at the problems that arose with the UEFI `pc` machines though. I will use the week while you are away to eventually test and debug a little. Enjoy your holiday!

**Comment by PaulYuuu** [comment]
Hello @pevogam , is `pc` + `OVMF` a valid usage? In fact, we can add it indeed to ensure compatibility, but I want to confirm with you this item if it's needed, thanks.

**Comment by pevogam** [comment]
> Hello @pevogam , is `pc` + `OVMF` a valid usage? In fact, we can add it indeed to ensure compatibility, but I want to confirm with you this item if it's needed, thanks.

Hi @PaulYuuu, I cannot speak from perspective of internal engineers but what I can say is that it used to work before which at least to me implies it is and should be supported.

**Comment by PaulYuuu** [comment]
> > Hello @pevogam , is `pc` + `OVMF` a valid usage? In fact, we can add it indeed to ensure compatibility, but I want to confirm with you this item if it's needed, thanks.
> 
> Hi @PaulYuuu, I cannot speak from perspective of internal engineers but what I can say is that it used to work before which at least to me implies it is and should be supported.

OK, let's keep the compatibility with `pc` machine, thanks.

**Comment by pevogam** [comment]
Alright, feel free to ping me once you push the changes. I could test them then.

**Comment by PaulYuuu** [comment]
> Alright, feel free to ping me once you push the changes. I could test them then.

Seems that nothing needs to be changed again? L1127 fixed the issue of `pc` + `ovmf`. Or are there other problems I missed?

**Comment by pevogam** [comment]
I had more troubles some of which I described back in the thread from https://github.com/avocado-framework/avocado-vt/pull/2777/files#r572382697. I will rerun and re-explain the issues here early next week :+1:

**Review by pevogam — APPROVED** [review]
Hi @PaulYuuu, I am running full tests on top of your branch and they are passing. I will do a few more hours of test running but it seems this fixes everything completely. LGTM

**Comment by pevogam** [comment]
Hmmm something is worse here, it turns out some tests passed because the OVMF variables weren't added. I get this qemu command line:
```
    -S  \
    -name 'mis4' \
    -machine pc  \
    -nodefaults \
    -device virtio-vga,bus=pci.0,addr=0x2 \
    -m 1024  \
    -smp 2,maxcpus=2,cores=1,threads=1,dies=1,sockets=2  \
    -cpu 'EPYC' \
    -chardev socket,nowait,server,path=/tmp/monitor-hmp1-20210301-161743-mUxD8XPW,id=hmp_id_hmp1  \
    -mon chardev=hmp_id_hmp1,mode=readline \
    -chardev socket,nowait,server,path=/tmp/monitor-catch_monitor-20210301-161743-mUxD8XPW,id=hmp_id_catch_monitor  \
    -mon chardev=hmp_id_catch_monitor,mode=readline \
    -chardev socket,nowait,server,path=/tmp/serial-serial0-20210301-161743-mUxD8XPW,id=chardev_serial0 \
    -device isa-serial,id=serial0,chardev=chardev_serial0 \
    -device piix3-usb-uhci,id=isdnhub,bus=pci.0,addr=0x3 \
    -device usb-tablet,id=usb-tablet1,bus=isdnhub.0,port=1 \
    -drive id=drive_image1,if=none,cache=none,format=qcow2,file=/mnt/local/images/c36_mis4/BusinessServer_64.qcow2 \
    -device virtio-blk-pci,id=image1,drive=drive_image1,bus=pci.0,addr=0x4 \
    -device virtio-net-pci,mac=02:00:00:00:04:01,id=idUjMjjN,netdev=idbq51HT,bus=pci.0,addr=0x5  \
    -netdev tap,id=idbq51HT,fd=28 \
    -device virtio-net-pci,mac=02:00:00:00:04:02,id=idNJlfJP,netdev=idEWmONc,bus=pci.0,addr=0x6  \
    -netdev tap,id=idEWmONc,fd=31 \
    -device virtio-net-pci,mac=02:00:00:00:04:03,id=id3rbT6f,netdev=idzAP2us,bus=pci.0,addr=0x7  \
    -netdev tap,id=idzAP2us,fd=8 \
    -drive cache=none,media=cdrom,file=/mnt/local/isos/business-server-6-9-0.iso,bus=0,unit=0  \
    -vnc :0  \
    -boot menu=off,order=cdn,once=dcn \
    -enable-kvm
```
despite having defined
```
16:17:42 DEBUG|     ovmf_code_filename = OVMF_CODE.fd
16:17:42 DEBUG|     ovmf_path = /usr/share/edk2/ovmf/
16:17:42 DEBUG|     ovmf_vars_filename = OVMF_VARS.fd
```
Am I missing something obvious?

File: virttest/qemu_devices/qcontainer.py
Comment (PaulYuuu) [inline virttest/qemu_devices/qcontainer.py:1208]: Sorry @pevogam ,I missing this line before, to extend pflash devices.
After this fix, I tried to boot a guest with `pc` machine using OVMF, the result has a bit different from you.
```
2021-03-02 00:12:55,287 test             L0324 DEBUG|     machine_type = pc
2021-03-02 00:12:55,293 test             L0324 DEBUG|     ovmf_code_filename = OVMF_CODE.secboot.fd
2021-03-02 00:12:55,293 test             L0324 DEBUG|     ovmf_path = /usr/share/OVMF
2021-03-02 00:12:55,294 test             L0324 DEBUG|     ovmf_vars_filename = OVMF_VARS.fd
```

```
MALLOC_PERTURB_=1  /usr/libexec/qemu-kvm \
    -S  \
    -name 'avocado-vt-vm1'  \
    -sandbox on  \
    -blockdev node-name=file_ovmf_code,driver=file,filename=/usr/share/OVMF/OVMF_CODE.secboot.fd,auto-read-only=on,discard=unmap \
    -blockdev node-name=drive_ovmf_code,driver=raw,read-only=on,file=file_ovmf_code \
    -blockdev node-name=file_ovmf_vars,driver=file,filename=/var/lib/avocado/data/avocado-vt/images/rhel830-64-virtio-scsi.qcow2_VARS.fd,auto-read-only=on,discard=unmap \
    -blockdev node-name=drive_ovmf_vars,driver=raw,read-only=off,file=file_ovmf_vars \
    -machine pc,pflash0=drive_ovmf_code,pflash1=drive_ovmf_vars  \
    -nodefaults \
    -device VGA,bus=pci.0,addr=0x2 \
    -m 15360  \
    -smp 20,maxcpus=20,cores=10,threads=1,dies=1,sockets=2  \
    -cpu 'IvyBridge',+kvm_pv_unhalt \
```

The machine command provided by you is `-machine pc`, it lost `pflash0=drive_ovmf_code,pflash1=drive_ovmf_vars`, this is strange and I am not sure if your test case dropped OVMF variables before creating the guest.

Comment (pevogam) [reply virttest/qemu_devices/qcontainer.py:1208]: After your recent push, it now correctly adds the variables here too:
```
15:29:35 INFO | Running qemu command (reformatted):
/usr/bin/qemu-kvm \
    -S  \
    -name 'mis4' \
    -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2/ovmf/OVMF_CODE.fd \
    -drive if=pflash,format=raw,file=/var/lib/avocado/data/avocado-vt/images/mis4_BusinessServer_64.qcow2_VARS.fd \
    -machine pc  \
    -nodefaults \
    -device virtio-vga,bus=pci.0,addr=0x2 \
    -m 1024  \
    -smp 2,maxcpus=2,cores=1,threads=1,dies=1,sockets=2  \
    -cpu 'EPYC' \
```
I still wonder why the variables are added right before the machine type and not together with the other drives but I guess the Qemu command line is not strictly ordered and this is not that important anyway.

Alright, so tests pass now. All we need is one more review and this can be merged :)

Comment (PaulYuuu) [reply virttest/qemu_devices/qcontainer.py:1208]: OK, `-drive` mode will not append `pflash0=drive_ovmf_code,pflash1=drive_ovmf_vars` in the `-machine pc`, I forgot it.

This is in order to align with libvirt, add these two pflash devices to generate machine extra params at first, pass them into machine command.
```
/usr/libexec/qemu-kvm \
-name guest=avocado-vt-vm1,debug-threads=on \
-S \
-object secret,id=masterKey0,format=raw,file=/var/lib/libvirt/qemu/domain-1-avocado-vt-vm1/master-key.aes \
-blockdev {"driver":"file","filename":"/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw","node-name":"libvirt-pflash0-storage","auto-read-only":true,"discard":"unmap"} \
-blockdev {"node-name":"libvirt-pflash0-format","read-only":true,"driver":"raw","file":"libvirt-pflash0-storage"} \
-blockdev {"driver":"file","filename":"/var/lib/libvirt/qemu/nvram/avocado-vt-vm1_VARS.fd","node-name":"libvirt-pflash1-storage","auto-read-only":true,"discard":"unmap"} \
-blockdev {"node-name":"libvirt-pflash1-format","read-only":false,"driver":"raw","file":"libvirt-pflash1-storage"} \
-machine virt-rhel8.4.0,accel=kvm,usb=off,dump-guest-core=off,gic-version=3,pflash0=libvirt-pflash0-format,pflash1=libvirt-pflash1-format,memory-backend=mach-virt.ram
```

File: virttest/qemu_devices/qcontainer.py
Comment (pevogam) [inline virttest/qemu_devices/qcontainer.py:1074]: Can this be implemented? For instance:
```diff
                 # FIXME: Put the pflash_var file and the image file in the same
                 # directory if necessary.
-                pflash_vars_path = img_name + "_VARS.fd"
+                pflash_vars_path = os.path.join(current_data_dir, img_name + "_VARS.fd")
                 pflash0, pflash1 = (firmware_name + '_code',
```
In the current implementation I keep getting inconsistent locations for the pflash var file. Is there a reason to postpone implementing this here?

**Review by pevogam — CHANGES_REQUESTED** [review]

File: virttest/qemu_devices/qcontainer.py
Comment (PaulYuuu) [reply virttest/qemu_devices/qcontainer.py:1074]: > In the current implementation I keep getting inconsistent locations for the pflash var file.

Really? pflash files should all in `data_dir.DATA_DIR/images`.

Some images are named similarly `image_name = images/f31-${vm_arch_name}`, so if you want to save fd files under `current_data_dir`, I think we can easily use
```
                path = storage.get_image_filename_filesytem(img_params,
                                                            current_data_dir)
                pflash_vars_path = path + "_VARS.fd"
```
To avoid fd files save under like `/var/lib/avocado/data/avocado-vt/` but not `/var/lib/avocado/data/avocado-vt/images/`.

Comment (pevogam) [reply virttest/qemu_devices/qcontainer.py:1074]: Either way is good, the current changes however don't achieve this and I end up with
```
    -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2/ovmf/OVMF_CODE.fd \
    -drive if=pflash,format=raw,file=BusinessServer_64.qcow2_VARS.fd \
```
which is not admissible. The fix I suggested above has been tested so far and it works. I can test your fix too if you want to push it. We should drop the `FIXME` and implement this now rather than later as it leads to undesirable behaviors like above.

Comment (PaulYuuu) [reply virttest/qemu_devices/qcontainer.py:1074]: https://github.com/avocado-framework/avocado-vt/blob/6142776eac1ea69f0b2461759fecf4899f91fb1f/virttest/qemu_devices/qcontainer.py#L1072-L1074

Currently, var files are under `DATA_DIR/images`, I am not sure why `-drive if=pflash,format=raw,file=BusinessServer_64.qcow2_VARS.fd \` will happen.

My test result:
```
    -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw \
    -drive if=pflash,format=raw,file=/var/lib/avocado/data/avocado-vt/images/avocado-vt-vm1_rhel830-aarch64-virtio-scsi.qcow2_VARS.fd \
```

Comment (pevogam) [reply virttest/qemu_devices/qcontainer.py:1074]: > Currently, var files are under `DATA_DIR/images`, I am not sure why `-drive if=pflash,format=raw,file=BusinessServer_64.qcow2_VARS.fd \` will happen.

We are using a custom `images_base_dir` like `images_base_dir = /mnt/local/images/c37_mis1/` which doesn't seem to take effect here.

Also there is no denying that the current implementation of `storage.get_image_filename_filesytem` and all related functions is very hard to read and keeps overwriting the path on each occasion. I am not sure which logical path is taken there but I am not sure it is worth investigating it instead of simply making sure to set the path in a simple manner at this stage.

Comment (PaulYuuu) [reply virttest/qemu_devices/qcontainer.py:1074]:
```
2021-03-04 07:40:29,291 test             L0326 DEBUG|     image_format = qcow2
2021-03-04 07:40:29,291 test             L0326 DEBUG|     image_name = image1
2021-03-04 07:40:29,292 test             L0326 DEBUG|     images = image1
2021-03-04 07:40:29,292 test             L0326 DEBUG|     images_base_dir = /mnt/local/images/c37_mis1/

    -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw \
    -drive if=pflash,format=raw,file=/var/lib/avocado/data/avocado-vt/images/avocado-vt-vm1_image1.qcow2_VARS.fd \
```

It seems that `BusinessServer_64.qcow2_VARS.fd` is generated by `pflash_vars_name = "%s_%s_VARS.fd" % (self.vmname, img_name)`, but the next line `pflash_vars_path = os.path.join(data_dir.DATA_DIR, "images", pflash_vars_name)` does not work.
Even if `data_dir.DATA_DIR` is empty, the pflash path should also be `images/BusinessServer_64.qcow2_VARS.fd`, so the result is strange.

Comment (pevogam) [reply virttest/qemu_devices/qcontainer.py:1074]: Yes, so just to summarize the desired behavior and if we both agree: the goal is to be able to define `images_base_dir = /some/path` and thus make the code aware that we want to place the VARS file in the images base directory. Do you agree? This doesn't seem to be achievable with the current changes so far.

Comment (PaulYuuu) [reply virttest/qemu_devices/qcontainer.py:1074]: Sure, I agree with you. This item is okay for me, it can even make the code more concise.

I will update it first and then wait for our results before making a decision.

**Comment by PaulYuuu** [comment]
Test results with the latest code:

- with `images_base_dir` and `-drive`:
    ```
    -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw \
    -drive if=pflash,format=raw,file=/mnt/local/images/c37_mis1/avocado-vt-vm1_image1.qcow2_VARS.fd \
    ```
- with `images_base_dir` and `-blockdev`:
    ```
    -blockdev node-name=file_aavmf_code,driver=file,filename=/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw,auto-read-only=on,discard=unmap \
    -blockdev node-name=drive_aavmf_code,driver=raw,read-only=on,file=file_aavmf_code \
    -blockdev node-name=file_aavmf_vars,driver=file,filename=/mnt/local/images/c37_mis1/avocado-vt-vm1_image1.qcow2_VARS.fd,auto-read-only=on,discard=unmap \
    -blockdev node-name=drive_aavmf_vars,driver=raw,read-only=off,file=file_aavmf_vars \
    ```
- without `images_base_dir` and `-drive`:
    ```
    -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw \
    -drive if=pflash,format=raw,file=/home/kvm_autotest_root/images/avocado-vt-vm1_rhel830-aarch64-virtio-scsi.qcow2_VARS.fd \
- without `images_base_dir` and `-blockdev`:
    ```
    -blockdev node-name=file_aavmf_code,driver=file,filename=/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw,auto-read-only=on,discard=unmap \
    -blockdev node-name=drive_aavmf_code,driver=raw,read-only=on,file=file_aavmf_code \
    -blockdev node-name=file_aavmf_vars,driver=file,filename=/home/kvm_autotest_root/images/avocado-vt-vm1_rhel830-aarch64-virtio-scsi.qcow2_VARS.fd,auto-read-only=on,discard=unmap \
    -blockdev node-name=drive_aavmf_vars,driver=raw,read-only=off,file=file_aavmf_vars \
    ```

**Comment by pevogam** [comment]
> Test results with the latest code:
> 
>     * with `images_base_dir` and `-drive`:
>       ```
>       -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw \
>       -drive if=pflash,format=raw,file=/mnt/local/images/c37_mis1/avocado-vt-vm1_image1.qcow2_VARS.fd \
>       ```
> 
>     * with `images_base_dir` and `-blockdev`:
>       ```
>       -blockdev node-name=file_aavmf_code,driver=file,filename=/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw,auto-read-only=on,discard=unmap \
>       -blockdev node-name=drive_aavmf_code,driver=raw,read-only=on,file=file_aavmf_code \
>       -blockdev node-name=file_aavmf_vars,driver=file,filename=/mnt/local/images/c37_mis1/avocado-vt-vm1_image1.qcow2_VARS.fd,auto-read-only=on,discard=unmap \
>       -blockdev node-name=drive_aavmf_vars,driver=raw,read-only=off,file=file_aavmf_vars \
>       ```
> 
>     * without `images_base_dir` and `-drive`:
>       ```
>       -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw \
>       -drive if=pflash,format=raw,file=/home/kvm_autotest_root/images/avocado-vt-vm1_rhel830-aarch64-virtio-scsi.qcow2_VARS.fd \
>       ```
> 
>     * without `images_base_dir` and `-blockdev`:
>       ```
>       -blockdev node-name=file_aavmf_code,driver=file,filename=/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw,auto-read-only=on,discard=unmap \
>       -blockdev node-name=drive_aavmf_code,driver=raw,read-only=on,file=file_aavmf_code \
>       -blockdev node-name=file_aavmf_vars,driver=file,filename=/home/kvm_autotest_root/images/avocado-vt-vm1_rhel830-aarch64-virtio-scsi.qcow2_VARS.fd,auto-read-only=on,discard=unmap \
>       -blockdev node-name=drive_aavmf_vars,driver=raw,read-only=off,file=file_aavmf_vars \
>       ```

Current fix works here too:
```
    -name 'mis4' \
    -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2/ovmf/OVMF_CODE.fd \
    -drive if=pflash,format=raw,file=/mnt/local/images/c38_mis4/mis4_BusinessServer_64.qcow2_VARS.fd \
    -machine q35 \
```
I will do some vm resetting, recreating, and reusing tests to confirm this one more time but I think we are nearly done.

**Review by pevogam — APPROVED** [review]
All my previous tests passed, this seems to be the fix.

**Comment by PaulYuuu** [comment]
Thank you, @pevogam .

@luckyh , could you help review again? Thanks in advance.

**Review by luckyh — APPROVED** [review]
ACK

**Comment by luckyh** [comment]
Thanks all.

**Pull request merged**
